lib/web/assistant.go

```diff
@@ -19,46 +19,785 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !enabled.Enabled {
+		return trace.AccessDenied("Assist is not enabled")
+	}
+
+	return nil
+}
+
+// runAssistant upgrades the HTTP connection to a websocket and starts a chat loop.
+func runAssistant(h *Handler, w http.ResponseWriter, r *http.Request,
+	sctx *SessionContext, site reversetunnel.RemoteSite,
+) error {
+	q := r.URL.Query()
+	conversationID := q.Get("conversation_id")
+	if conversationID == "" {
+		return trace.BadParameter("conversation ID is required")
+	}
+
+	authClient, err := sctx.GetClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := checkAssistEnabled(authClient, r.Context()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	identity, err := createIdentityContext(sctx.GetUser(), sctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	ctx, err := h.cfg.SessionControl.AcquireSessionContext(r.Context(), identity, h.cfg.ProxyWebAddr.Addr, r.RemoteAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	authAccessPoint, err := site.CachingAccessPoint()
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to get auth access point.")
+		return trace.Wrap(err)
+	}
+
+	netConfig, err := authAccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		h.log.WithError(err).Debug("Unable to fetch cluster networking config.")
+		return trace.Wrap(err)
+	}
+
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+		CheckOrigin:     func(r *http.Request) bool { return true },
+	}
+
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		errMsg := "Error upgrading to websocket"
+		h.log.WithError(err).Error(errMsg)
+		http.Error(w, errMsg, http.StatusInternalServerError)
+		return nil
+	}
+
+	// Note: This time should be longer than OpenAI response time.
+	keepAliveInterval := netConfig.GetKeepAliveInterval()
+	err = ws.SetReadDeadline(deadlineForInterval(keepAliveInterval))
+	if err != nil {
+		h.log.WithError(err).Error("Error setting websocket readline")
+		return nil
+	}
+	defer ws.Close()
+
+	// Update the read deadline upon receiving a pong message.
+	ws.SetPongHandler(func(_ string) error {
+		ws.SetReadDeadline(deadlineForInterval(keepAliveInterval))
+		return nil
+	})
+
+	ws.SetCloseHandler(func(code int, text string) error {
+		h.log.Warnf("closing assistant websocket: %v %v", code, text)
+		return nil
+	})
+
+	go startPingLoop(ctx, ws, keepAliveInterval, h.log, nil)
+
+	client, err := getAssistantClient(ctx, h.cfg.ProxyClient,
+		h.cfg.ProxySettings, h.cfg.OpenAIConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	chat := client.NewChat(sctx.cfg.User)
+
+	// existing conversation, retrieve old messages
+	messages, err := authClient.GetAssistantMessages(ctx, &assist.GetAssistantMessagesRequest{
+		ConversationId: conversationID,
+		Username:       sctx.GetUser(),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// restore conversation context.
+	for _, msg := range messages.GetMessages() {
```

Comment:
pretty much all of this is racy across tabs iirc but I haven't done enough frontend work to be sure. should we open an issue for post-release fix to prevent issues when the user opens multiple assist tabs?

Reply:
Good point. GH has even a button for that https://github.com/gravitational/teleport/issues/26170